File: algorithms_and_datastructures/merge_sort.py
def divide(arr):
    midpoint = len(arr) // 2
    left = arr[:midpoint]
    right = arr[midpoint:]
    return left, right



def merge(left, right):
    arr = []
    i = 0
    j = 0

    while(i < len(left) and j < len(right)):
        if left[i] < right[j]:
            arr.append(left[i])
            i += 1
        else:
            arr.append(right[j])
            j+= 1

    while(i < len(left)):
        arr.append(left[i])
        i += 1
    while(j < len(right)):
        arr.append(right[j])
        j += 1
    return arr




def merge_sort(arr):
    #Base case
    if len(arr) <= 1:
        return arr

    #Divide recursively until hitting base case
    left, right = divide(arr)
    left = merge_sort(left)
    right = merge_sort(right)

    #Conquer (Merge elements)
    arr2 = merge(left, right)

    return arr2

File: algorithms_and_datastructures/merge_sort.js
function divide(array){
    //Runs on O(logn) time + the time taken by the slice method 1q@
    let mid = Math.floor(array.length / 2);
    let left = array.slice(0, mid);
    let right = array.slice(mid);
    return [left, right];
}

function merge(left, right){
    //runs in O(n) time
    let arr = [];
    while(left.length && right.length){
        if(left[0] < right[0]){
            arr.push(left.shift());
        }else{
            arr.push(right.shift());
        }
    }
    return ([...arr, ...left, ...right]);

}

function mergeSort(array){
    if(array.length <= 1){
        return array;
    }
    let divided = divide(array);
    let left = mergeSort(divided[0]);
    let right = mergeSort(divided[1]);
    return merge(left, right);
}

function verify(array){
    if(array.length <= 1){
        return true;
    }
    return array[0] < array[1] && verify(array.slice(1));
}

let a1 = [3, 20, 100, 10, 5, 1, 8, 30];
let a2 = mergeSort(a1);

console.log(verify(a1));
console.log(verify(a2));